runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java

```diff
@@ -154,112 +159,169 @@ private static long debugDuration(long nowMs, long startMs) {
+          if (isShutdown) {
+            break;
+          }
+          debugMetrics.recordStart();
+          streamClosed = false;
+          requestObserver.reset();
           onNewStream();
-          if (clientClosed.get()) {
+          if (clientClosed) {
             halfClose();
           }
           return;
         }
+      } catch (WindmillStreamShutdownException e) {
+        logger.debug("Stream was shutdown waiting to start.", e);
       } catch (Exception e) {
-        LOG.error("Failed to create new stream, retrying: ", e);
+        logger.error("Failed to create new stream, retrying: ", e);
         try {
           long sleep = backoff.nextBackOffMillis();
-          sleepUntil.set(Instant.now().getMillis() + sleep);
-          Thread.sleep(sleep);
-        } catch (InterruptedException | IOException i) {
+          debugMetrics.recordSleep(sleep);
+          sleeper.sleep(sleep);
+        } catch (InterruptedException ie) {
+          Thread.currentThread().interrupt();
+          logger.info(
+              "Interrupted during {} creation backoff. The stream will not be created.",
+              getClass());
+          break;
+        } catch (IOException ioe) {
           // Keep trying to create the stream.
         }
       }
     }
+
+    // We were never able to start the stream, remove it from the stream registry. Otherwise, it is
+    // removed when closed.
+    streamRegistry.remove(this);
   }
 
-  protected final Executor executor() {
-    return executor;
+  /**
+   * Execute the runnable using the {@link #executor} handling the executor being in a shutdown
+   * state.
+   */
+  protected final void executeSafely(Runnable runnable) {
+    try {
+      executor.execute(runnable);
+    } catch (RejectedExecutionException e) {
+      logger.debug("{}-{} has been shutdown.", getClass(), backendWorkerToken);
+    }
   }
 
-  public final synchronized void maybeSendHealthCheck(Instant lastSendThreshold) {
-    if (lastSendTimeMs.get() < lastSendThreshold.getMillis() && !clientClosed.get()) {
+  public final void maybeSendHealthCheck(Instant lastSendThreshold) {
+    if (!clientClosed && debugMetrics.lastSendTimeMs() < lastSendThreshold.getMillis()) {
       try {
         sendHealthCheck();
       } catch (RuntimeException e) {
-        LOG.debug("Received exception sending health check.", e);
+        logger.debug("Received exception sending health check.", e);
       }
     }
   }
 
   protected abstract void sendHealthCheck();
 
-  // Care is taken that synchronization on this is unnecessary for all status page information.
-  // Blocking sends are made beneath this stream object's lock which could block status page
-  // rendering.
+  /**
+   * @implNote Care is taken that synchronization on this is unnecessary for all status page
+   *     information. Blocking sends are made beneath this stream object's lock which could block
+   *     status page rendering.
+   */
+  @SuppressWarnings("GuardedBy")
```

Comment:
why does this need to be suppressed? seems better to fix it

it can cause TSAN failures even if it is benign from VM memory model.

Reply:
it means we will need to acquire "this" lock here

Reply:
made streamClosed a volatile like clientClosed